py/src/nupic/optimization/ae.py

```diff
@@ -0,0 +1,898 @@
+experiment which generated it.  If the experiment is "foo/bar.py" then AE's
+directory is "foo/bar_ae/".  The primary output of this program is the file
+"foo/bar_ae/lab_report.txt" which contains a summary of its operations.  The lab
+report format is:
+
+1) Introduction.  This text is not parsed, it is preserved so you can keep notes
+here  This area is initialized with hopefully useful information, including
+the experiment name, timestamps.
+
+2) Methods.  This section contains the default parameters and the command line
+invocation.
+
+3) Summary of each experiment.  Each experiments summary contains the following
+information:
+    1) Modified Parameters & their New Value. This is the only required field,
+    the remaining fields will be generated and written to this file as soon as
+    the report is loaded.  You may choose to manually add experiments to this
+    lab report in this way.
+    2) MD5 Checksum of Parameters and Command Line.  This hash checksum is used
+    to uniquely identify an experimental setup, it's the name of the
+    experiment.  These hashes are used in filenames and searching for a hash
+    finds all references to it.
+    3) File Path of Experiment Journal
+    4) Number of Attempted Runs
+    5) Score of each Completed Run
+    6) Mean & Standard Deviation of Scores
+    7) Notes, these are not parsed they are preserved so you can keep notes here
+
+This program keeps a Journal of each experimental setup.  Journals are named
+after the hash of the parameters & command line they contain, with the extension
+".journal".  Journals start with a self contained description of how to
+reproduce the experiment, followed by a section for every run of this
+experiment.  The section for each run contains the output (std-out & std-err) of
+the program, as well as diagnostic information such as timestamps and memory
+usage reports.  Files with the extension ".log" are temporary files for in-
+progress experiment, and when the experiment finishes running they are copied to
+their journal and then the ".log" file is deleted.
+"""
+
+# TODO: Default parameters need better handling...
+
+# TODO: ExperimentSummary should X-Ref with journal file.  Journal should
+# contain complete parameter set, needed for full automatic reproducibility.
+# Changing default parameters should not make experiments inaccessible.
+
+# TODO: Maybe the command line invocation should be included in the experiment
+# hash?  Then I could experiment with the CLI args within a single lab report.
+
+# TODO: Every run should track elapsed time and report the average in the
+# experiment journal & summary.  Some of these experiments clearly take 4x
+# longer than others but its not recorded.
+
+# TODO: Log files should have memory usage ...
+
+import argparse
+import os
+import sys
+import shutil
+import random
+import pprint
+import time
+import datetime
+import tempfile
+import multiprocessing
+import resource
+import signal # Probably not X-platform ...
```

Comment:
would this be a problem? but hey..worst case, the feature will be avail only for certain OSs

Reply:
I looked this up, This use of `signal` can be replaced with multithreading.timer which _is_ X-platform.